runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptionsTest.java

```diff
@@ -177,9 +177,9 @@ public void testDefaultInvalidGcpTempLocation() {
   @Test
   public void testDefaultStagingLocationUnset() {
     DataflowPipelineOptions options = PipelineOptionsFactory.as(DataflowPipelineOptions.class);
+    options.setProject("");
     thrown.expect(IllegalArgumentException.class);
-    thrown.expectMessage(
-        "Error constructing default value for stagingLocation: gcpTempLocation is missing.");
+    thrown.expectMessage("--project is a required option.");
     options.getStagingLocation();
   }
```

Comment:
We're missing unit tests of the new code.

I think this would probably have highlighted the issue we identified yesterday. Specifically, I expect this configuration to fail today, but be fixable. We should add this test (and others, I'll let @tgroh guide) to ensure we get it right and make sure it continues to work.

* tempLocation = something-no-gs
* stagingLocation = missing
* gcsTempLocation = missing
* options.getGcpTempLocation() -- will fail rather than getting a default value, because StagingLocationFactory would not have been invoked.